From a7ee433c9fd11c7156640a05207bee1bc08abd5c Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Tue, 18 Feb 2025 03:50:33 -0800
Subject: [PATCH 15/18] ddraw: Use signed types for GetAvailableVidMem.

Otherwise applications see negative memory if the true amount is >INT_MAX.
---
 dlls/ddraw/ddraw.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/dlls/ddraw/ddraw.c b/dlls/ddraw/ddraw.c
index c4ce81234bc..ad91fd18135 100644
--- a/dlls/ddraw/ddraw.c
+++ b/dlls/ddraw/ddraw.c
@@ -2011,7 +2011,8 @@ static HRESULT WINAPI ddraw7_GetAvailableVidMem(IDirectDraw7 *iface, DDSCAPS2 *c
     if (free)
     {
         free_vidmem = wined3d_device_get_available_texture_mem(ddraw->wined3d_device);
-        *free = framebuffer_size > free_vidmem ? 0 : free_vidmem - framebuffer_size;
+        free_vidmem = min(INT_MAX, free_vidmem);
+        *free = (INT)framebuffer_size > (INT)free_vidmem ? 0 : (INT)(free_vidmem - framebuffer_size);
         TRACE("Free video memory %#lx.\n", *free);
     }
 
@@ -2020,8 +2021,8 @@ static HRESULT WINAPI ddraw7_GetAvailableVidMem(IDirectDraw7 *iface, DDSCAPS2 *c
         struct wined3d_adapter_identifier desc = {0};
 
         hr = wined3d_adapter_get_identifier(ddraw->wined3d_adapter, 0, &desc);
-        total_vidmem = min(UINT_MAX, desc.video_memory);
-        *total = framebuffer_size > total_vidmem ? 0 : total_vidmem - framebuffer_size;
+        total_vidmem = min(INT_MAX, desc.video_memory);
+        *total = (INT)framebuffer_size > (INT)total_vidmem ? 0 : (INT)(total_vidmem - framebuffer_size);
         TRACE("Total video memory %#lx.\n", *total);
     }
 
-- 
2.53.0

